crates/valence_command/src/command_scopes.rs

```diff
@@ -0,0 +1,324 @@
+//! Scope graph for the Valence Command system.
+//!
+//! ## Breakdown
+//! Each scope is a node in the graph. a path from one node to another indicates
+//! that the first scope implies the second. A colon in the scope name indicates
+//! a sub-scope. you can use this to create a hierarchy of scopes. for example,
+//! the scope "valence:command" implies "valence:command:tp". this means that if
+//! a player has the "valence:command" scope, they can use the "tp" command.
+//!
+//! You may also link scopes together in the registry. this is useful for admin
+//! scope umbrellas. for example, if the scope "valence:admin" is linked to
+//! "valence:command:teleport", It means that if a player has the
+//! "valence:admin" scope, they can use the "teleport" command.
+//!
+//! # Example
+//! ```
+//! use valence_command::command_scopes::{CommandScopeRegistry, Scope};
+//!
+//! let mut registry = CommandScopeRegistry::new();
+//!
+//! // add a scope to the registry
+//! registry.add_scope("valence:command:teleport");
+//!
+//! // we added 4 scopes to the registry. "valence", "valence:command", "valence:command:teleport",
+//! // and the root scope.
+//! assert_eq!(registry.scope_count(), 4);
+//!
+//! registry.add_scope("valence:admin");
+//!
+//! // add a scope to the registry with a link to another scope
+//! registry.link("valence:admin", "valence:command:teleport");
+//!
+//! // the "valence:admin" scope implies the "valence:command:teleport" scope
+//! assert_eq!(
+//!     registry.grants("valence:admin", "valence:command:teleport"),
+//!     true
+//! );
+//! ```
+
+use std::collections::HashMap;
+use std::fmt::{Debug, Display, Formatter};
+
+use bevy_ecs::prelude::Component;
+use bevy_ecs::system::Resource;
+use petgraph::dot;
+use petgraph::dot::Dot;
+use petgraph::prelude::*;
+
+/// Store the scope graph and provide methods for querying it.
+#[derive(Clone, Resource)]
+pub struct CommandScopeRegistry {
+    graph: Graph<Scope, ()>,
+    string_to_node: HashMap<String, NodeIndex>,
+    root: NodeIndex,
+}
+
+impl Debug for CommandScopeRegistry {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{:?}",
+            Dot::with_config(&self.graph, &[dot::Config::EdgeNoLabel])
+        )?;
+        Ok(())
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Ord, PartialOrd)]
+pub struct Scope {
+    pub name: String,
+}
```

Comment:
```suggestion
pub struct Scope(pub String);
```

Reply:
Ditto, I prefer to keep it expanded as it keeps code more readable then using .0 all the time.